sqlmesh/core/engine_adapter/spark.py

```diff
@@ -40,12 +61,29 @@ def _insert_overwrite_by_condition(
         where: t.Optional[exp.Condition] = None,
         columns_to_types: t.Optional[t.Dict[str, exp.DataType]] = None,
     ) -> None:
-        if isinstance(query_or_df, (pd.DataFrame, PySparkDataFrame)):
-            self._insert_pyspark_df(
-                table_name, self._ensure_pyspark_df(query_or_df), overwrite=True
-            )
+        table = exp.to_table(table_name)
+        df = self.try_get_pandas_df(query_or_df)
+        pyspark_df = self.try_get_pyspark_df(query_or_df)
+        if self._use_spark_session and (df is not None or pyspark_df):
+            if df is not None:
+                pyspark_df = self._ensure_pyspark_df(df)
+            assert pyspark_df
+            self._insert_pyspark_df(table_name, pyspark_df, overwrite=True)
         else:
-            super()._insert_overwrite_by_condition(table_name, query_or_df, where, columns_to_types)
+            if df is not None:
```

Comment:
Why are we not falling back to parent class here?

Reply:
Base class assumes you can't do an `INSERT OVERWRITE` so it does a transactional delete + insert append.